src/Storages/StorageMemory.cpp

```diff
@@ -123,13 +128,18 @@ Pipes StorageMemory::read(
 
     for (size_t stream = 0; stream < num_streams; ++stream)
     {
-        BlocksList::iterator begin = data.begin();
-        BlocksList::iterator end = data.begin();
+        BlocksList::iterator first = data.begin();
+        BlocksList::iterator last = data.begin();
 
-        std::advance(begin, stream * size / num_streams);
-        std::advance(end, (stream + 1) * size / num_streams);
+        std::advance(first, stream * size / num_streams);
+        std::advance(last, (stream + 1) * size / num_streams);
 
-        pipes.emplace_back(std::make_shared<MemorySource>(column_names, begin, end, *this, metadata_snapshot));
+        if (first == last)
```

Comment:
I do not think this case is possible at all:
 - loop body is only executed for `num_streams > 0`
 - Assume `R = size / num_streams`. `size >= num_streams` implies that `R >= 1`
 -  `stream + 1 > stream` implies `last > first` for any integral R >= 1

Reply:
I suppose you are right. But I wouldn't like calculating it reading this code.
This check costs nothing.